pkg/storage/fs/eos/upload.go

```diff
@@ -0,0 +1,319 @@
+	log.Debug().Interface("info", info).Msg("eos: NewUpload")
+
+	fn := info.MetaData["filename"]
+	if fn == "" {
+		return nil, errors.New("eos: missing filename in metadata")
+	}
+	info.MetaData["filename"] = filepath.Clean(info.MetaData["filename"])
+
+	dir := info.MetaData["dir"]
+	if dir == "" {
+		return nil, errors.New("eos: missing dir in metadata")
+	}
+	info.MetaData["dir"] = filepath.Clean(info.MetaData["dir"])
+
+	log.Debug().Interface("info", info).Msg("eos: resolved filename")
+
+	info.ID = uuid.New().String()
+
+	binPath, err := fs.getUploadPath(ctx, info.ID)
+	if err != nil {
+		return nil, errors.Wrap(err, "eos: error resolving upload path")
+	}
+	user, err := getUser(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "eos: no user in ctx")
+	}
+	info.Storage = map[string]string{
+		"Type":     "EOSStore",
+		"Username": user.Username,
+	}
+	// Create binary file with no content
+
+	file, err := os.OpenFile(binPath, os.O_CREATE|os.O_WRONLY, defaultFilePerm)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	u := &fileUpload{
+		info:     info,
+		binPath:  binPath,
+		infoPath: binPath + ".info",
+		fs:       fs,
+	}
+
+	// writeInfo creates the file by itself if necessary
+	err = u.writeInfo()
+	if err != nil {
+		return nil, err
+	}
+
+	return u, nil
+}
+
+// TODO use a subdirectory in the shadow tree
+func (fs *eosfs) getUploadPath(ctx context.Context, uploadID string) (string, error) {
+	return filepath.Join(fs.conf.CacheDirectory, uploadID), nil
+}
+
+// GetUpload returns the Upload for the given upload id
+func (fs *eosfs) GetUpload(ctx context.Context, id string) (tusd.Upload, error) {
+	binPath, err := fs.getUploadPath(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+	infoPath := binPath + ".info"
+	info := tusd.FileInfo{}
+	data, err := ioutil.ReadFile(infoPath)
+	if err != nil {
+		return nil, err
+	}
+	if err := json.Unmarshal(data, &info); err != nil {
+		return nil, err
+	}
+
+	stat, err := os.Stat(binPath)
+	if err != nil {
+		return nil, err
+	}
+
+	info.Offset = stat.Size()
+
+	return &fileUpload{
+		info:     info,
+		binPath:  binPath,
+		infoPath: infoPath,
+		fs:       fs,
+	}, nil
+}
+
+type fileUpload struct {
+	// info stores the current information about the upload
+	info tusd.FileInfo
+	// infoPath is the path to the .info file
+	infoPath string
+	// binPath is the path to the binary file (which has no extension)
+	binPath string
+	// only fs knows how to handle metadata and versions
+	fs *eosfs
+}
+
+// GetInfo returns the FileInfo
+func (upload *fileUpload) GetInfo(ctx context.Context) (tusd.FileInfo, error) {
+	return upload.info, nil
+}
+
+// GetReader returns an io.Readerfor the upload
+func (upload *fileUpload) GetReader(ctx context.Context) (io.Reader, error) {
+	return os.Open(upload.binPath)
+}
+
+// WriteChunk writes the stream from the reader to the given offset of the upload
+// TODO use the grpc api to directly stream to a temporary uploads location in the eos shadow tree
```

Comment:
Send chunk to EOS directly

Reply:
tracked in https://github.com/cs3org/reva/issues/700